[WebXR] WPT WebXR/xrWebGLLayer_opaque_framebuffer_stencil incorrectly uses glReadPixels

djg opened the pull request:
#### 12fdd3513b715778f22b2fd3a4fc2d4fc4789d20
<pre>
[WebXR] WPT WebXR/xrWebGLLayer_opaque_framebuffer_stencil incorrectly uses glReadPixels
<a href="https://bugs.webkit.org/show_bug.cgi?id=301409">https://bugs.webkit.org/show_bug.cgi?id=301409</a>
<a href="https://rdar.apple.com/163318566">rdar://163318566</a>

Reviewed by Kimmo Kinnunen.

The OpenGL specification says only two combinations of format and type for any
internalformat are accepted in most cases. The first varies depending on the
internalformat of the currently bound rendering surface. For the format and
types used in the test, the combination format RGBA and type UNSIGNED_BYTE is
accepted. The second implementation-chosen format and type for this internalformat
may be determined by calling GetIntegerv with the symbolic constants
IMPLEMENTATION_COLOR_READ_FORMAT and IMPLEMENTATION_COLOR_READ_TYPE, which the
test is not doing.

Changing the format passed to glReadPixels from GL_RGB to GL_RBGA fixes the
test and follows the specification.

Renable and update expectations for xrWebGLLayer_opaque_framebuffer_stencil.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/302321@main">https://commits.webkit.org/302321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15ad8aaca7d712bcaf6a7fb79ede21567cbb5dcf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80002 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9a387dfb-af07-4704-9339-f0f9f3d733bb) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/800 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/729 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97890 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65794 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c5b37df8-3f31-4789-85d9-c307553313ad) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131537 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115210 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78506 "2 api tests failed or timed out") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5dfb9156-18fa-49bd-9f6c-3652a6e539bd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33318 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79261 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33800 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138430 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/655 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106425 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/732 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106245 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27078 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/577 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30080 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53030 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/744 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/616 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/675 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/695 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https.html:
```diff
@@ -183,7 +183,7 @@
     pixels[0] = pixels[1] = pixels[2] = pixels[3] = 30;
-    gl.readPixels(xrViewport.x + xrViewport.width / 2, xrViewport.y + xrViewport.height/4, 1, 1, gl.RGB, gl.UNSIGNED_BYTE, pixels);
+    gl.readPixels(xrViewport.x + xrViewport.width / 2, xrViewport.y + xrViewport.height/4, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
     if (pixels[0] == 0x0 && pixels[1] == 0xFF && pixels[2] == 0x0) { // green?
```
Also test and report the alpha.

kkinnunen-apple commented on LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https.html:
```diff
@@ -194,7 +194,7 @@
+    gl.readPixels(xrViewport.x + xrViewport.width / 2, xrViewport.y + xrViewport.height/2, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
     if (pixels[0] == 0xFF && pixels[1] == 0xFF && pixels[2] == 0xFF) { // white?
       // PASSED.
     } else if (pixels[0] == 0x0 && pixels[1] == 0xFF && pixels[2] == 0x0) { // green?
```
also test and report the alpha

kkinnunen-apple reviewed (approved):
Looks good, maybe add the alpha test to be consistent. (RGBA read of .alpha==false FBO results in alpha==0xFF)

djg committed:
[WebXR] WPT WebXR/xrWebGLLayer_opaque_framebuffer_stencil incorrectly uses glReadPixels
https://bugs.webkit.org/show_bug.cgi?id=301409
rdar://163318566

Reviewed by Kimmo Kinnunen.

The OpenGL specification says only two combinations of format and type for any
internalformat are accepted in most cases. The first varies depending on the
internalformat of the currently bound rendering surface. For the format and
types used in the test, the combination format RGBA and type UNSIGNED_BYTE is
accepted. The second implementation-chosen format and type for this internalformat
may be determined by calling GetIntegerv with the symbolic constants
IMPLEMENTATION_COLOR_READ_FORMAT and IMPLEMENTATION_COLOR_READ_TYPE, which the
test is not doing.

Changing the format passed to glReadPixels from GL_RGB to GL_RBGA fixes the
test and follows the specification.

Renable and update expectations for xrWebGLLayer_opaque_framebuffer_stencil.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer_stencil.https.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Canonical link: https://commits.webkit.org/302321@main
